hgeometry/src/Data/Geometry/Polygon/Convex.hs

```diff
@@ -383,7 +386,43 @@ minkowskiSum p q = ConvexPolygon . unsafeFromPoints $ merge' (f p) (f q)
     merge' _         _         = error "minkowskiSum: Should not happen"
 
 
-
+--------------------------------------------------------------------------------
+-- inConvex
+
+-- 1. Check if p is on left edge or right edge.
+-- 2. Do binary search:
+--       Find the largest n where p is on the right of 0 to n.
+-- 3. Check if p is on segment n,n+1
+-- 4. Check if p is in triangle 0,n,n+1
+
+-- | \( O(\log n) \)
+--   Check if a point lies inside a convex polygon, on the boundary, or outside of the
+--   convex polygon.
+inConvex :: forall p r. (Fractional r, Ord r)
+         => Point 2 r -> ConvexPolygon p r
+         -> PointLocationResult
+inConvex p (ConvexPolygon poly)
+  | onSegment p leftEdge  = OnBoundary
+  | onSegment p rightEdge = OnBoundary
+  | otherwise             = worker 1 n
+  where
+    p'        = p :+ undefined
+    n         = size poly - 1
+    point0    = point 0
+    leftEdge  = OpenLineSegment point0 (point n)
+    rightEdge = OpenLineSegment point0 (point 1)
```

Comment:
Are you sure these should be open line segments rather than closed ones? (Admittedly, I only recently (in the visibility polygon stuff) fixed that onSegment actually properly handles (partially) open segments as well).

Reply:
Pretty sure. I'm not trying to find a unique edge that contains the point. Just checking if the point is on /any/ edge.

Reply:
I think that if p is one of the vertices of the polygon treating the segments as Open won't work. Since 'onSegment' that is in master treats any segment (no matter what it says) as a ClosedLineSegment this has not triggered any failures yet, but in the visibility-polygon branch I've actually fixed that.

Reply:
There's a QC property that checks if the convex polygon vertices are considered to be OnBoundary. It hasn't failed so far.

`onSegment` appears to treat all segments as open rather than closed:

> λ> onSegment (Point2 0 0) (OpenLineSegment (ext $ Point2 0 0) (ext $ Point2 1 (1::Rational)))
> True
> λ> onSegment (Point2 1 1) (OpenLineSegment (ext $ Point2 0 0) (ext $ Point2 1 (1::Rational)))
> True
> λ> onSegment (Point2 0 0) (ClosedLineSegment (ext $ Point2 0 0) (ext $ Point2 1 (1::Rational)))
> True
> λ> onSegment (Point2 1 1) (ClosedLineSegment (ext $ Point2 0 0) (ext $ Point2 1 (1::Rational)))
> True

Reply:
Closed segments should include the endpoints; open segments should not. So the first two tests you listed are incorrect (in master; not in the visibility polygon branch). (Although the docs of onSegment did state that it would treat everything as closed)

Reply:
Ah, yeah then I want closed segments. :)